### ISSUE TITLE
Misdetected cycle between :extractIncludeTestProto and :compileKotlin

### ISSUE THREAD

**Issue by javiyt** [google/protobuf-gradle-plugin#624]
After upgrading to version 0.9.1 when we run `./gradlew build` task inside a Kotlin project we get the following error:
```
Misdetected cycle between :extractIncludeTestProto and :compileKotlin
```

Here is our plugins definition in the build.gradle.kts file:
```
plugins {
    val kotlinVersion = "1.7.20"
    kotlin("jvm") version kotlinVersion
    kotlin("plugin.spring") version kotlinVersion
    kotlin("kapt") version kotlinVersion

    id("idea")
    id("jacoco")
    id("application")

    id("com.github.ben-manes.versions") version "0.42.0" // https://plugins.gradle.org/plugin/com.github.ben-manes
    // .versions
    id("com.google.cloud.tools.jib") version "3.3.0" // https://plugins.gradle.org/plugin/com.google.cloud.tools.jib
    id("com.google.protobuf") version "0.9.1" // https://plugins.gradle.org/plugin/com.google.protobuf

    id("io.gitlab.arturbosch.detekt") version "1.21.0" // https://plugins.gradle.org/plugin/io.gitlab.arturbosch.detekt

    id("nebula.resolution-rules") version "9.0.0" // https://plugins.gradle.org/plugin/nebula.resolution-rules

    id("org.ajoberstar.grgit") version "5.0.0" // https://plugins.gradle.org/plugin/org.ajoberstar.grgit
    id("org.flywaydb.flyway") version "9.4.0" // https://plugins.gradle.org/plugin/org.flywaydb.flyway
    id("org.jlleitschuh.gradle.ktlint") version "11.0.0" // https://plugins.gradle.org/plugin/org.jlleitschuh.gradle.ktlint
    id("org.owasp.dependencycheck") version "7.2.1" // https://plugins.gradle.org/plugin/org.owasp.dependencycheck
    id("org.sonarqube") version "3.4.0.2513" // https://plugins.gradle.org/plugin/org.sonarqube
    id("org.springframework.boot") version "2.7.4" // https://plugins.gradle.org/plugin/org.springframework.boot
}
```

Any help to fix it would be appreciated.

**Comment by rougsig** [comment]
Could you please provide minimal reprodusible project?

**Comment by rougsig** [comment]
@javiyt 

Could you describe your project configuration in more detail?
1. Do you modify the protobuf tasks by yourself? Not from the `generateProtoTasks` block.
2. Do you modify gradle configurations? (create, remove, etc)
3. How do you provide proto files to the project? (raw *.proto files, from any archive, etc)
4. Could you please provide full stacktrace? (just run build with `--stacktrace` flag)
5. Does project have any `testProtobuf` configuration dependencies?

**Comment by javiyt** [comment]
Hello @rougsig as far as it's a private project I can't share it, I'm trying to create another project with the same conditions as the one that is failing that I can share with you.
I can answer some of your questions:
1. We haven't modify any task.
2. We have some tweaks to improve kotlin compilation
3. From an archive
4. Let me check if I can share it
5. No, we don't use testProtobuf

**Comment by Spikhalskiy** [comment]
Face a similar problem. After upgrading from 0.8.19 to 0.9.1 getting
```
Circular dependency between the following tasks:
:temporal-serviceclient:classes
\--- :temporal-serviceclient:spotlessApply
     \--- :temporal-serviceclient:spotlessKotlinApply
          \--- :temporal-serviceclient:spotlessKotlin
               \--- :temporal-serviceclient:generateTestProto
                    \--- :temporal-serviceclient:extractIncludeTestProto
                         \--- :temporal-serviceclient:classes (*)
```

`classes -> spotlessApply` link is created explicitly by
```
    classes.dependsOn 'spotlessApply'
```

On 0.8.19 the project builds fine.

**Comment by rougsig** [comment]
@Spikhalskiy What is the spotless version?

**Comment by Spikhalskiy** [comment]
> What is the spotless version?
6.11.0 with googleJavaFormat('1.15.0') and ktlint('0.47.1')

> Is it created by a flawless plugin or by you?
This dependency is setup in the project (our own) gradle build script. And it's been working with 0.8.x versions of this plugin fine.

**Comment by samzurcher** [comment]
When trying to upgrade from 0.8.18 to 0.9.1, we are seeing a similar issue:
```
Circular dependency between the following tasks:
:server:infrastructure:base:generateProto
\--- :server:infrastructure:base:kspKotlin
     \--- :server:infrastructure:base:generateProto (*)

(*) - details omitted (listed previously)
```
With 0.8.18 it works like a charm, only change is an upgrade from 0.8.18 to 0.9.1.

**Comment by rougsig** [comment]
@samzurcher Hello. Do you add ksp plugin and it'a all? No extra build script configuration?

**Comment by samzurcher** [comment]
We do have a rather complex build, use code in buildSrc and the KSP plugin to process our code. But all is fine with our build with version 0.8.19 but changing to 0.9.1 results in the above gradle problem.

**Comment by ejona86** [comment]
>     classes.dependsOn 'spotlessApply'

I have a very bad feeling about this. I think there's something very fundamentally wrong.

spotlessApply is not a task accepting input and producing output. Instead, it seems it would be rewriting source files in-place. That invalidates Gradle's dependency system and is bad except when run in a separate gradle invocation on non-generated source code. Even with fixing the circular dependency, I expect there will need to be further spotless configuration to exclude the generated code.

I think it is inherently broken to do `classes.dependsOn 'spotlessApply'` as you will have inconsistent builds. I think this is a broken usage of `spotlessApply`.

**Comment by Spikhalskiy** [comment]
> I have a very bad feeling about this. I think there's something very fundamentally wrong.

We want to force linting on every build. 

> it would be rewriting source files in-place

correct

> That invalidates Gradle's dependency system and is bad except when run in a separate gradle invocation on non-generated source code.

This may be ideologically wrong, but it serves the purpose and it works fine until 0.9.x version of this plugin.

> I think it is inherently broken to do classes.dependsOn 'spotlessApply' as you will have inconsistent builds. I think this is a broken usage of spotlessApply.

Honestly, for the majority of users of any build system, it's not important if something is "pure or not pure" from the point of view of the build system. Build systems come and go. If it works and serves the purpose - it's good enough.
If this usage is "not pure" and will not be supported by this plugin anymore, we will just stuck to the old version of the plugin for as long as we can, because "inconsistent builds" and "broken usage" lead to the exact outcome we want.

**Comment by ejona86** [comment]
> This may be ideologically wrong, but it serves the purpose and it works fine until 0.9.x version of this plugin.

No, not an ideological thing. It means your build is unsound. You are compiling .java files at the same time those files are rewritten. Whether it "works fine" could depend on things like "how many cores does your machine have" or "whether Chrome is running" or "how an object computes its hash code."

> We want to force linting on every build before classes are compiled.

For a build without any generated code, adding the dependency to `javaCompile` instead of `classes` would be a better start. But _just_ linting and not rewriting is probably necessary to make it have defined results.

**Comment by Spikhalskiy** [comment]
@ejona86 See for example this issue and discussion in spotless: https://github.com/diffplug/spotless/issues/453
Which shows that `compileJava.dependsOn('spotlessApply')` is a normal supported by maintainers usage of spotless.

Which also works with 0.8.x versions of this plugin and broken with 0.9.x with

```
Circular dependency between the following tasks:
:temporal-sdk:compileJava
\--- :temporal-serviceclient:compileJava
     \--- :temporal-serviceclient:spotlessApply
          \--- :temporal-serviceclient:spotlessKotlinApply
               \--- :temporal-serviceclient:spotlessKotlin
                    \--- :temporal-serviceclient:generateTestProto
                         \--- :temporal-serviceclient:extractIncludeTestProto
                              +--- :temporal-sdk:compileJava (*)
                              +--- :temporal-serviceclient:classes
                              |    \--- :temporal-serviceclient:compileJava (*)
                              +--- :temporal-serviceclient:compileJava (*)
                              +--- :temporal-test-server:compileJava
                              |    +--- :temporal-sdk:compileJava (*)
                              |    +--- :temporal-serviceclient:compileJava (*)
                              |    +--- :temporal-test-server:generateProto
                              |    |    \--- :temporal-test-server:extractIncludeProto
                              |    |         +--- :temporal-sdk:compileJava (*)
                              |    |         \--- :temporal-serviceclient:compileJava (*)
                              |    \--- :temporal-test-server:spotlessApply
                              |         \--- :temporal-test-server:spotlessKotlinApply
                              |              \--- :temporal-test-server:spotlessKotlin
                              |                   +--- :temporal-test-server:generateProto (*)
                              |                   \--- :temporal-test-server:generateTestProto
                              |                        \--- :temporal-test-server:extractIncludeTestProto
                              |                             +--- :temporal-sdk:compileJava (*)
                              |                             +--- :temporal-serviceclient:compileJava (*)
                              |                             +--- :temporal-test-server:classes
                              |                             |    \--- :temporal-test-server:compileJava (*)
                              |                             +--- :temporal-test-server:compileJava (*)
                              |                             \--- :temporal-testing:compileJava
                              |                                  +--- :temporal-sdk:compileJava (*)
                              |                                  +--- :temporal-serviceclient:compileJava (*)
                              |                                  \--- :temporal-test-server:compileJava (*)
                              \--- :temporal-testing:compileJava (*)
```

**Comment by Spikhalskiy** [comment]
Thank you for the explanation. In our project, we disable spotless for generated code anyway by doing `targetExclude '**/generated/*`.
It looks like the best course of action for us is to stay on an older version of protobuf plugin for now until spotless maintainers maybe find a way around this change.

**Comment by ejona86** [comment]
@samzurcher, do you know why your kspKotlin depends on generateProto? I played around with the [playground example](https://github.com/google/ksp/tree/main/examples/playground) a wasn't able to get such a dependency naturally.

**Comment by schmist** [comment]
@ejona86, we set the dependency `generateProto` -> `kspKotlin` explicitly since we generate proto files with ksp, but we do not know where the dependency `kspKotlin` -> `generateProto` is comming from. It was not there with version 0.8.x.

**Comment by ejona86** [comment]
Fixed by #631
